### PR TITLE
Associate WebGL textures with texture units

### PR DESCRIPTION
**Pull request by MortimerGoro**
Currently `bound_texture_2d` and `bound_texture_cube_map` fields are used to restore texture states and to get the current bindings in GetParams(...). But as soon as active texture is changed all the values can become dirty, leading to broken textures in some demos or invalid getParam(...) call results.

This PR implements the texture binding association with the texture units.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18580)

### CODE REVIEW

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -1938,12 +2001,32 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
     }
 
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8
+    #[allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: Why is this necessary?

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -110,6 +110,26 @@ macro_rules! object_binding_to_js_or_null {
...
+            rooted!(in($cx) let mut rval = NullValue());
+            if let Some(object) = $binding {
+                object.to_jsval($cx, rval.handle_mut());
+            }
+            rval.get()
+        }
+    };
+}
+
+macro_rules! bound_texture {
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: Rather than a macro, I think a `bound_texture` method that accepts an enum argument would be more readable.

**Review by jdm — CHANGES_REQUESTED** [review]
Looking for!

File: components/script/dom/webglrenderingcontext.rs (hunk 2 above)
Comment (MortimerGoro) [reply components/script/dom/webglrenderingcontext.rs]: Done

File: components/script/dom/webglrenderingcontext.rs (hunk 1 above)
Comment (MortimerGoro) [reply components/script/dom/webglrenderingcontext.rs]: Because of the dictionary iterator()
```  
Expression of type std::collections::hash_map::Iter<'_, u32, dom::webglrenderingcontext::TextureUnitBindings> must be rooted'''
```
Is there any better way to avoid that?

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -227,18 +274,36 @@ impl WebGLRenderingContext {
...
-    pub fn bound_texture_for_target(&self, target: &TexImageTarget) -> Option<DomRoot<WebGLTexture>> {
-        match *target {
-            TexImageTarget::Texture2D => self.bound_texture_2d.get(),
-            TexImageTarget::CubeMapPositiveX |
-            TexImageTarget::CubeMapNegativeX |
-            TexImageTarget::CubeMapPositiveY |
-            TexImageTarget::CubeMapNegativeY |
-            TexImageTarget::CubeMapPositiveZ |
-            TexImageTarget::CubeMapNegativeZ => self.bound_texture_cube_map.get(),
+    fn bound_texture(&self, target: u32) -> Option<DomRoot<WebGLTexture>> {
```
Comment (jdm) [inline components/script/dom/webglrenderingcontext.rs]: If we take a custom enum argument that only accepts Texture2d and TextureCubeMap, we don't need to return an Option. This will simplify other calling code, and then we can remove the optional_root_object_tojs_or_null macro.

File: components/script/dom/webglrenderingcontext.rs (hunk 1 above)
Comment (jdm) [reply components/script/dom/webglrenderingcontext.rs]: Add a new entry to https://github.com/servo/servo/blob/master/components/script_plugins/unrooted_must_root.rs#L57.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/webglrenderingcontext.rs (hunk 3 above)
Comment (MortimerGoro) [reply components/script/dom/webglrenderingcontext.rs]: Ok, I'll create the new enum. The result will be an Option too because the bound texture is nullable.

Comment (jdm) [reply components/script/dom/webglrenderingcontext.rs]: Oh, I see. In that case I don't think there's any strong benefit to using a separate enum.

**Comment by MortimerGoro** [comment]
Changes made, thanks for the unrooted_must_root info!

**Review by jdm — APPROVED** [review]
